usr/lib/cca_stdll/cca_specific.c

```diff
@@ -5754,6 +5836,59 @@ CK_RV token_specific_aes_cbc(STDLL_TokData_t * tokdata,
     return rc;
 }
 
+/**
+ * This routine is currently only used when the operation is performed using
+ * a protected key. Therefore we don't have (and don't need) an cca
+ * fallback here.
+ */
+CK_RV token_specific_aes_xts(STDLL_TokData_t *tokdata, SESSION *session,
+                             CK_BYTE *in_data, CK_ULONG in_data_len,
+                             CK_BYTE *out_data, CK_ULONG *out_data_len,
+                             OBJECT *key_obj, CK_BYTE *init_v,
+                             CK_BBOOL encrypt, CK_BBOOL initial,
+                             CK_BBOOL final, CK_BYTE *iv)
+{
+    CK_ATTRIBUTE *attr = NULL;
+    CK_RV rc;
+#ifndef NO_PKEY
+    CK_MECHANISM mech = { CKM_AES_XTS, NULL, 0 };
+#endif
+
+#ifdef NO_PKEY
+    UNUSED(session);
+#endif
+
+    if (((struct cca_private_data *)tokdata->private_data)->inconsistent) {
+        TRACE_ERROR("%s\n", ock_err(ERR_DEVICE_ERROR));
+        return CKR_DEVICE_ERROR;
+    }
+
+    rc = template_attribute_get_non_empty(key_obj->template, CKA_IBM_OPAQUE, &attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Could not find CKA_IBM_OPAQUE for the key.\n");
+        return rc;
+    }
+
+#ifndef NO_PKEY
+    /* CCA token protected key option */
+    rc = ccatok_pkey_check(tokdata, session, key_obj, &mech);
+    switch (rc) {
+    case CKR_OK:
+        rc = pkey_aes_xts(key_obj, init_v, in_data, in_data_len,
+                          out_data, out_data_len, encrypt, initial, final, iv);
+        goto done;
+    default:
+        goto done;
+    }
+#endif /* NO_PKEY */
+
+#ifndef NO_PKEY
+done:
```

Comment:
We could move the done label into the #ifndef NO_PKEY block above and save this one